src/python/pants/pantsd/pailgun_server.py

```diff
@@ -189,10 +197,23 @@ def process_request_thread(self, request, client_address):
     # Instantiate the request handler.
     Native().override_thread_logging_destination_to_just_pantsd()
     handler = self.RequestHandlerClass(request, client_address, self)
+
+
+    with self.handling_request_lock:
```

Comment:
I think that rather than doing this here, you could do it inside of an override of `process_request_thread`, which is the entry point for each request thread, and which is called from within `process_request`.

And I think that rather than exiting, you could have a thread running a request hold this lock, and a thread that is interested in running a request use `if self.handling_request_lock.acquire(false):` to attempt to acquire, and then log something either once or in a loop until it acquires successfully.

Reply:
So I'm not sure I agree that we want to wait until we acquire the lock successfully, because many runs are potentially long, and blocking until it's finished can lead to bad UX.

This is not a problem if we log to the client, but I'm concerned about how this would impact metrics. If we start `./pants --enable-pantsd compile ::` on one side, and immediatly start `./pants --enable-pantsd list src/scala::` on the other, and the second run blocks until the first is done, the timing will be way off.

Modifying the metrics to account for this is possible, but probably out of the scope of this PR.

Reply:
I'm also not sure I understand the first paragraph, this code is already inside a `process_request_thread`.

Reply:
I don't think failing is a good user experience though... especially since it's something we want to support in the future.

And re: `process_request_thread`: whoops! Quite so. Misread and thought that this was located elsewhere.

Reply:
One concrete reason is that failing in automated usecases (IntelliJ) would require clients to add retries, or for humans to retry if their IDE is busy refreshing.

Reply:
I have implemented blocking (will turn into an option next), which I guess is fine since we definitely want to optionize the timeout time, and a possible value could be 0 (I don't want to wait).

Thanks for the insight!

Reply:
I'm going to resolve this conversation to clean things up for the real actual review.